### PR TITLE
Adding `#pragma CPROVER enable "named-check"`

### PR DESCRIPTION
**Pull request by remi-delmas-3000**
This PR adds the possibility to selectively activate checks :
- on blocks of C code using `#pragma CPROVER enable "named-check"`
- to individual GOTO instructions and expressions using `source_location.add_pragma("enable:named-check")`

It is possible to enable/disable a check, to push a level on the stack and override the check status using another pragma enable/disable pragma.
Attempts to enable and disable a same check at the same stack level are detected and cause a PARSING_ERROR in the C front end and a INVARIANT violation in `goto_check`.

Modifications:
- the collection of allocations on the whole goto program using `goto_checkt::collect_allocation` is now always on
- a `messaget log` attribute is added to `goto_checkt` to be able to log warnings in case a same check is enabled/disabled more than once on a given GOTO instruction
- the `flag_resett` class is extended to detect multiple 'set_flag' operations on a same flag
- the ansi-c lexer/parser is modified to accept `#pragma CPROVER enable "named check"` and flag shadowing
- tests are added in `regression/cbmc/pragma_cprover_enable*` to test the feature

This feature is needed as part of the function and loop contracts verification, to be able to selectively activate checks on  instructions used for assigns-clause checking, in order to guarantee the soundness of the analysis, even if these checks are not globally activated in the current or ulterior `goto-instrument` or `cbmc` passes

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/analyses/goto_check.cpp
Comment (SaswatPadhi) [inline src/analyses/goto_check.cpp]: It seems like we are setting this to `true` in the constructor and no other method sets it to `false`, so wouldn't this be _always_ `true`? And if that's the case, can we completely remove this and also remove the `if(!... & !...) return;` check in `collect_allocations`?

Or did I miss a code path where it won't be `true`?

File: src/analyses/goto_check.cpp
Comment (SaswatPadhi) [inline src/analyses/goto_check.cpp]: I think the error message could be improved a bit. Having "more than one CPROVER enable/disable check" is fine as long as they are _different_ checks.
I think we want to say that we _the same check_ being enabled/disabled multiple times times at a goto instruction.

File: regression/cbmc/pragma_cprover_enable3/test.desc
Comment (SaswatPadhi) [inline regression/cbmc/pragma_cprover_enable3/test.desc]: [nit] missing newline. Also in a few other files.

File: regression/cbmc/pragma_cprover_enable1/main.c
Comment (SaswatPadhi) [inline regression/cbmc/pragma_cprover_enable1/main.c]:
```suggestion
  // pop all checks added since the last push
```

[nit] also in other test files

**Review by SaswatPadhi — COMMENTED** [review]
Nice :) Mostly minor comments below.

Can we add at least one regression test to check nesting of both disable and enable pragmas, something like:

```c
  int n;

#pragma CPROVER check push
#pragma CPROVER check enable "signed-overflow"
   // generate overflow checks for the following addition
  int x = n + n;
#pragma CPROVER check push
#pragma CPROVER check disable "signed-overflow"
   // do not generate overflow checks for the following increment operations
  ++n;
  n++;
#pragma CPROVER check pop
   // generate overflow checks for the following addition
  int b = n + n;
#pragma CPROVER check pop
```

File: doc/cprover-manual/properties.md
Comment (feliperodri) [inline doc/cprover-manual/properties.md]:
```
(to re-enable all properties disabled or enabled before or since the last `#pragma CPROVER check push`)
```
This wording seems confusing. Could you re-write it?

File: doc/cprover-manual/properties.md
Comment (feliperodri) [inline doc/cprover-manual/properties.md]: Same as above.

File: regression/cbmc/pragma_cprover_enable1/main.c
Comment (feliperodri) [reply regression/cbmc/pragma_cprover_enable1/main.c]: +1

File: regression/cbmc/pragma_cprover_enable1/test.desc
Comment (feliperodri) [inline regression/cbmc/pragma_cprover_enable1/test.desc]: Add newline.

File: src/analyses/goto_check.cpp
Comment (feliperodri) [reply src/analyses/goto_check.cpp]: +1

**Review by feliperodri — APPROVED** [review]
Only minor comments (modulo Saswat's comments).

**Comment by chrisr-diffblue** [comment]
Just a general git housekeeping comment - it's generally better to use `git pull --rebase origin develop` (replace `origin` with whatever you've called your remote - `origin` being the default if you didn't specify/change the remote name) to update your branch with the latest `develop` branch, rather than doing a merge of `develop` - it will lead to cleaner commits and cleaner git history.

File: src/goto-instrument/goto_instrument_parse_options.cpp
Comment (tautschnig) [inline src/goto-instrument/goto_instrument_parse_options.cpp:1335]: Use `ui_message_handler` here (should be the same as `log.get_message_handler()`, but is more consistent with the other code over here.

File: src/ansi-c/scanner.l
Comment (tautschnig) [inline src/ansi-c/scanner.l:427]: Unnecessary removal of blank line.

File: src/ansi-c/ansi_c_parser.h
Comment (tautschnig) [inline src/ansi-c/ansi_c_parser.h]: "Returns true iff the same" (missing "iff")

File: src/ansi-c/ansi_c_parser.h
Comment (tautschnig) [inline src/ansi-c/ansi_c_parser.h]: Unnecessary whitespace change?

File: src/ansi-c/ansi_c_parser.h
Comment (tautschnig) [inline src/ansi-c/ansi_c_parser.h]: Whitespace change?

File: src/ansi-c/ansi_c_parser.h
Comment (tautschnig) [inline src/ansi-c/ansi_c_parser.h]: Whitespace change?

File: src/ansi-c/ansi_c_parser.h
Comment (tautschnig) [inline src/ansi-c/ansi_c_parser.h]: Whitespace change?

File: src/ansi-c/ansi_c_parser.h
Comment (tautschnig) [inline src/ansi-c/ansi_c_parser.h]: Whitespace change?

File: src/ansi-c/ansi_c_parser.h
Comment (tautschnig) [inline src/ansi-c/ansi_c_parser.h]: Whitespace change?

File: src/ansi-c/ansi_c_parser.cpp
Comment (tautschnig) [inline src/ansi-c/ansi_c_parser.cpp]: Whitespace change?

File: src/ansi-c/ansi_c_parser.cpp
Comment (tautschnig) [inline src/ansi-c/ansi_c_parser.cpp]: Whitespace change?

File: src/ansi-c/ansi_c_parser.cpp
Comment (tautschnig) [inline src/ansi-c/ansi_c_parser.cpp]: Whitespace change?

File: jbmc/src/jbmc/jbmc_parse_options.cpp
Comment (tautschnig) [inline jbmc/src/jbmc/jbmc_parse_options.cpp]: Use `ui_message_handler`

File: jbmc/src/janalyzer/janalyzer_parse_options.cpp
Comment (tautschnig) [inline jbmc/src/janalyzer/janalyzer_parse_options.cpp]: Use `ui_message_handler`

File: jbmc/src/jdiff/jdiff_parse_options.cpp
Comment (tautschnig) [inline jbmc/src/jdiff/jdiff_parse_options.cpp]: Use `ui_message_handler`

File: src/analyses/goto_check.cpp
Comment (tautschnig) [inline src/analyses/goto_check.cpp]: Removal of blank line?

File: src/analyses/goto_check.cpp
Comment (tautschnig) [inline src/analyses/goto_check.cpp]: Shouldn't this invariant just be maintained within `set_flag`? It seems it is checked every time you use `set_flag`.

File: src/ansi-c/ansi_c_parser.cpp
Comment (tautschnig) [inline src/ansi-c/ansi_c_parser.cpp]: Whitespace changes?

File: src/ansi-c/ansi_c_parser.cpp
Comment (tautschnig) [inline src/ansi-c/ansi_c_parser.cpp]: Please review all changes in this file for whitespace-only changes.

**Review by tautschnig — CHANGES_REQUESTED** [review]
Functionality looks ok, but I think this PR needs some cleanup to remove as much noise as possible.

File: src/analyses/goto_check.cpp
Comment (remi-delmas-3000) [reply src/analyses/goto_check.cpp]: this variable was removed and collection is now always on

File: src/analyses/goto_check.cpp
Comment (remi-delmas-3000) [reply src/analyses/goto_check.cpp]: message updated to include the name of the check

File: regression/cbmc/pragma_cprover_enable3/test.desc
Comment (remi-delmas-3000) [reply regression/cbmc/pragma_cprover_enable3/test.desc]: fixed

File: regression/cbmc/pragma_cprover_enable1/main.c
Comment (remi-delmas-3000) [reply regression/cbmc/pragma_cprover_enable1/main.c]: comment removed

File: doc/cprover-manual/properties.md
Comment (remi-delmas-3000) [reply doc/cprover-manual/properties.md]: documentation updated

File: doc/cprover-manual/properties.md
Comment (remi-delmas-3000) [reply doc/cprover-manual/properties.md]: documentation updated

File: jbmc/src/jbmc/jbmc_parse_options.cpp
Comment (remi-delmas-3000) [reply jbmc/src/jbmc/jbmc_parse_options.cpp]: corrected

File: jbmc/src/janalyzer/janalyzer_parse_options.cpp
Comment (remi-delmas-3000) [reply jbmc/src/janalyzer/janalyzer_parse_options.cpp]: corrected

File: jbmc/src/jdiff/jdiff_parse_options.cpp
Comment (remi-delmas-3000) [reply jbmc/src/jdiff/jdiff_parse_options.cpp]: corrected

File: src/analyses/goto_check.cpp
Comment (remi-delmas-3000) [reply src/analyses/goto_check.cpp]: corrected

File: src/analyses/goto_check.cpp
Comment (remi-delmas-3000) [reply src/analyses/goto_check.cpp]: The invariant is now maintained inside `flag_resett`. To be able to generate readable error messages`flag_resett::set_flag` now takes the flag name and instruction as parameters.

File: src/goto-instrument/goto_instrument_parse_options.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/goto_instrument_parse_options.cpp:1335]: corrected

File: src/ansi-c/ansi_c_parser.h
Comment (SaswatPadhi) [inline src/ansi-c/ansi_c_parser.h:13]: Do we still need the `#include <set>` below?

File: src/ansi-c/ansi_c_parser.h
Comment (SaswatPadhi) [inline src/ansi-c/ansi_c_parser.h:146]: I see why you have introduced these functions although `pragma_cprover_stack` is `public`.

Can we make it `private` / `protected` now? Just so we don't allow messing with `pragma_cprover_stack` directly and use these functions instead?

File: src/ansi-c/ansi_c_parser.cpp
Comment (SaswatPadhi) [inline src/ansi-c/ansi_c_parser.cpp:225]:
```suggestion
    auto check_name = id2string(pragma.first);
```

File: src/ansi-c/ansi_c_parser.cpp
Comment (SaswatPadhi) [inline src/ansi-c/ansi_c_parser.cpp]: Since `check_name` gets `std::string` (not `const char *`) type from the `auto` definition above, we can use the overloaded `+` in C++ to simplify this expression:

```suggestion
    auto full_name = (pragma.second ? "enable:" : "disable:") + check_name;
```

File: src/analyses/goto_check.cpp
Comment (SaswatPadhi) [inline src/analyses/goto_check.cpp]: I am not sure if I understand why `err_msg` needs to be a lambda function. It seems like it's called immediately (at the `INVARIANT` check) within the `set_flag` call. Can we not just pass a `string` here instead?

File: src/analyses/goto_check.cpp
Comment (SaswatPadhi) [inline src/analyses/goto_check.cpp]: Same comment as above.
Can we not just pass a `string` here instead?

File: src/analyses/goto_check.cpp
Comment (SaswatPadhi) [inline src/analyses/goto_check.cpp]: Hmm, does this function require a mutable reference? I can see why `set_flag` does, but I don't think `seen` does? How about:

```suggestion
  bool seen(const bool &flag)
```

File: src/analyses/goto_check.cpp
Comment (SaswatPadhi) [inline src/analyses/goto_check.cpp]: Hmm, we seem to be taking the address of `flag` here. So if we have two `flag`s with the same content but different addresses then this would break?

Do we have some mechanism to guard against it? Or if there's some other reason as to why this is a non-issue, we should mention that in the comments above this function.

**Review by SaswatPadhi — CHANGES_REQUESTED** [review]
A few more comments; nothing major

File: src/analyses/goto_check.cpp
Comment (remi-delmas-3000) [reply src/analyses/goto_check.cpp]: the class `flag_resett` is file-local so as long as the discipline is respected in this file it should be ok.

File: src/ansi-c/ansi_c_parser.cpp
Comment (remi-delmas-3000) [reply src/ansi-c/ansi_c_parser.cpp]: modif added

File: src/analyses/goto_check.cpp
Comment (remi-delmas-3000) [reply src/analyses/goto_check.cpp]: this lambda is used to defer the creation of the error string until it is actually needed (if ever).

File: src/analyses/goto_check.cpp
Comment (remi-delmas-3000) [reply src/analyses/goto_check.cpp:2048]: This lambda allows to defer the creation of the string until it is needed

File: src/analyses/goto_check.cpp
Comment (tautschnig) [inline src/analyses/goto_check.cpp]: I might be wrong, but I believe you can shorten this to `flags_to_reset.insert({&flag, flag});` with C++11.

File: src/analyses/goto_check.cpp
Comment (tautschnig) [inline src/analyses/goto_check.cpp]: Seems redundant?

File: src/analyses/goto_check.cpp
Comment (tautschnig) [inline src/analyses/goto_check.cpp]: Not that it matters a whole lot, but the parameter type should be `const bool *` to make clear that no modification will occur. Also, the method should be marked `const`.

File: src/analyses/goto_check.cpp
Comment (tautschnig) [inline src/analyses/goto_check.cpp]: Given that every call requires the instruction to be passed (and the instruction is always the same), should the constructor already take that?

File: src/ansi-c/ansi_c_parser.cpp
Comment (tautschnig) [inline src/ansi-c/ansi_c_parser.cpp]: I'm wondering whether `pragma_cprover_stack.emplace_back({});` would actually work (and be more concise)?

File: src/ansi-c/ansi_c_parser.cpp
Comment (tautschnig) [inline src/ansi-c/ansi_c_parser.cpp]: Use `const irep_idt &`

File: src/ansi-c/ansi_c_parser.cpp
Comment (tautschnig) [inline src/ansi-c/ansi_c_parser.cpp]: `const irep_idt &`

**Review by tautschnig — APPROVED** [review]

File: src/analyses/goto_check.cpp
Comment (remi-delmas-3000) [reply src/analyses/goto_check.cpp]: now passing flag name and instructions to set_flag

File: src/analyses/goto_check.cpp
Comment (SaswatPadhi) [reply src/analyses/goto_check.cpp]: Or just `flags_to_reset.emplace(&flag, flag);` to avoid creating a temporary pair.

File: src/ansi-c/ansi_c_parser.cpp
Comment (SaswatPadhi) [inline src/ansi-c/ansi_c_parser.cpp:227]:
```suggestion
    auto check_name = id2string(pragma.first);
    auto full_name =
      (pragma.second ? "enable:" : "disable:") + check_name;
```

**Review by SaswatPadhi — APPROVED** [review]
Approving modulo Michael's remaining comments and one minor suggestion:

File: src/ansi-c/ansi_c_parser.cpp
Comment (remi-delmas-3000) [reply src/ansi-c/ansi_c_parser.cpp:227]: `scanner.l ` uses `std::string` everywhere so I'll keep this as is for consistency

File: src/analyses/goto_check.cpp
Comment (remi-delmas-3000) [reply src/analyses/goto_check.cpp]: fixed

File: src/ansi-c/ansi_c_parser.cpp
Comment (SaswatPadhi) [reply src/ansi-c/ansi_c_parser.cpp:227]: Yes, but not sure why it has to be kept this way here though ... Using `auto` is not only more succinct but also makes it future proof (disconnects this from the underlying type, which is base `std::string` today)

File: src/analyses/goto_check.cpp
Comment (remi-delmas-3000) [reply src/analyses/goto_check.cpp]: fixed

File: src/ansi-c/ansi_c_parser.cpp
Comment (remi-delmas-3000) [reply src/ansi-c/ansi_c_parser.cpp]: seems to work with `pragma_cprover_stack.emplace_back()`

File: src/analyses/goto_check.cpp
Comment (remi-delmas-3000) [reply src/analyses/goto_check.cpp]: method removed

File: src/ansi-c/ansi_c_parser.cpp
Comment (remi-delmas-3000) [reply src/ansi-c/ansi_c_parser.cpp]: fixed

File: src/ansi-c/ansi_c_parser.cpp
Comment (remi-delmas-3000) [reply src/ansi-c/ansi_c_parser.cpp]: fixed

File: src/ansi-c/ansi_c_parser.cpp
Comment (remi-delmas-3000) [reply src/ansi-c/ansi_c_parser.cpp:227]: this would only be patching two lines out of several dozen that could use `auto`  the whole lexer would need refactoring to be effectively future proof.

**Comment by feliperodri** [comment]
@peterschrammel @martin-cs could you take a look at this PR? It's missing code owner approval.

File: src/ansi-c/ansi_c_parser.cpp
Comment (SaswatPadhi) [reply src/ansi-c/ansi_c_parser.cpp:227]: Up to you 🙂

**Comment by martin-cs** [comment]
@feliperodri Apologies for the lag; I missed this one.

**Comment by martin-cs** [comment]
This would be easier to review if it were split up into multiple commits.  Parsing, implementation, testing would seem a reasonable start.

**Comment by remi-delmas-3000** [comment]
@martin-cs @feliperodri @tautschnig I reworked the commit history it should be good for the final review now

**Review by peterschrammel — APPROVED** [review]

**Comment by martin-cs** [comment]
Thank you @remi-delmas-3000 it makes reviewing so much easier!

**Pull request merged**
